### DISCUSSION TITLE
how do i add a countdown for each maped component??

### DISCUSSION THREAD

**Discussion by PhilipZi** [neuefische/questions#115] [Web] [answered]
Guten Tag zusammen, 

ich würde gerne wissen, wie ich es umsetzten kann, einen CountDownTimer zu implementieren, der bei einer ersten gempapptem Komponente den value runterzählt bis dieser auf null ist und dann automatisch auf die nächste Komponente springt um dort den generierten value abzuspielen? 
Klingt bestimmt komplizierter als es ist, aber die logik ist schon nicht einfach würde ich behaupten. 
Siehe Bild, in welcher Folge die countdowns abgespielt werden sollen:
![Screenshot 2022-08-26 at 16 03 52](https://user-images.githubusercontent.com/102238180/186921568-91eca8d9-9d50-4985-a3f6-7db80ec183b8.png)

ich habe einen counter entwickelt weiss aber leider nicht wie und wo ich diesen richtig anwenden soll: siehe Bild für meine Counterfunction: 
![Screenshot 2022-08-26 at 16 04 57](https://user-images.githubusercontent.com/102238180/186921803-dfcd91a0-5895-4b96-a1f1-ba14c7b4feeb.png)

der Code ist dieser wo der Counter eingebaut werden soll: 
![Screenshot 2022-08-26 at 16 06 13](https://user-images.githubusercontent.com/102238180/186922042-4dd8cef3-2b97-471d-a443-fb2ca74f1c96.png)

Hier der Link zu meinem Github, ich weiss, dass es bestimmt schwierig ist zu verstehen, was ich möchte also feel free to ask, ich hoffe ich kann euch genauer erklären was ich möchte und bin sehr froh über Hilfe. 

https://github.com/PhilipZi/capstone-project/pull/19

**Comment by mpagels** [comment]
Ich würde es so machen: Du hast ein `State` der alle deine `Exercises` in einem Array enthält und auf deiner `/countdown` Page angezeigt wird. 
Das könnte in etwa so aussehen:

```js
[
  { 
    id: 0,
    name: "Squads",
    timer: 500, // <- seconds
  },
  { 
    id: 1,
    name: "Bizeps",
    timer: 300,
  },
  { 
    id: 2,
    name: "Abs",
    timer: 100,
  },
]
```

Und die Logik ist dann, dass du erst den `timer` runterzählst von den Squads, und wenn der `timer` auf 0 ist, dann zählst du den `timer` von den Bizeps runter. Usw. Beispiel: https://l14lq2.csb.app/

Als Tipp: Prototype mal in Codesanbox, schreibe ein Countdown, wo du erstmal eine Exercise runterzählst. Ohne das Ganze drumherum. Stumpf ein Countdown. Dann füge eine zweite Exercise hinzu und probiere die Logik anzupassen, dass wenn eine Exercise auf 0 ist, er die andere Exercise nehmen soll.

**Reply by PhilipZi** [reply]
danke!

**Reply by PhilipZi** [reply]
Das Beispiel ist sehr cool und danke für die Ansicht! Aber ich weiss nicht genau ob das in meinem Fall geht, da ich ja immer variierende Summen and exercises habe, die exercises sind immer exercises und variieren nicht im Namen aber sie variieren in der Rheinfolge und in der Summe wie viele ablaufen. Kann ich denn in diesem Array die Values von timer über probs aus dem Usestore übergeben? geht das?

**Comment by PhilipZi** [comment, marked as answer]
Ich habe es wie folgt gelöst :
````
const timers = useMemo(() => {
		const _sets = [];
		for (let sets = 0; sets < currentExercise.sets; sets++) {
			for (let repition = 0; repition < currentExercise.repetition; repition++) {
				_sets.push({
					variant: 'Exercise',
					minutes: currentExercise.exercise.minutes,
					seconds: currentExercise.exercise.seconds,
				});
				if (
					repition < currentExercise.repetition - 1 &&
					(currentExercise.pause.minutes > 0 || currentExercise.pause.seconds > 0)
				) {
					_sets.push({
						variant: 'Pause',
						minutes: currentExercise.pause.minutes,
						seconds: currentExercise.pause.seconds,
					});
				}
			}
			if (
				sets < currentExercise.sets - 1 &&
				(currentExercise.setPause.minutes > 0 || currentExercise.setPause.seconds > 0)
			) {
				_sets.push({
					variant: 'Setpause',
					minutes: currentExercise.setPause.minutes,
					seconds: currentExercise.setPause.seconds,
				});
			}
		}
		return _sets;
	}, [
		currentExercise.exercise.minutes,
		currentExercise.exercise.seconds,
		currentExercise.pause.minutes,
		currentExercise.pause.seconds,
		currentExercise.repetition,
		currentExercise.setPause.minutes,
		currentExercise.setPause.seconds,
		currentExercise.sets,
	]);

	return (
		<StyledLayout>
			<StyledBorder />

			<StyledHeader variant="counter">
				<MoveBackButton />
				<StyledH1>
					<StyledDumbbellIcon>
						<DumbbellIcon />
					</StyledDumbbellIcon>
					Your Workout
					<StyledDumbbellIcon variant="rotated">
						<DumbbellIcon />
					</StyledDumbbellIcon>
				</StyledH1>
			</StyledHeader>

			<StyledTimerCardsContainer>
				{timers.map((timer, idx) => (
					<TimerCard
						key={idx}
						variant={timer.variant}
						minutes={timer.minutes}
						seconds={timer.seconds}
						running={currentIndex === idx}
						onFinish={() => {
							setCurrentIndex(currentIndex + 1);
						}}
					/>
				))}
			</StyledTimerCardsContainer>
			<NavigationBar />
		</StyledLayout>
	);
}
`````

`````
export default function TimerCard({variant, minutes, seconds, running, onFinish}) {
	const [_minutes, setMinutes] = useState(minutes);
	const [_seconds, setSeconds] = useState(seconds);

	const timerOn = useStore(state => state.timerOn);

	useEffect(() => {
		if (!running) return;
		const timer = setInterval(() => {
			if (timerOn) {
				if (_minutes === 0 && _seconds === 0) {
					onFinish();
				} else if (_seconds === 0) {
					setMinutes(_minutes - 1);
					setSeconds(59);
				} else {
					setSeconds(_seconds - 1);
				}
			} else {
				clearInterval(timer);
			}
		}, 1000);
		return () => clearInterval(timer);
	}, [_minutes, _seconds, running, onFinish, timerOn]);

	return (
		<StyledTimerCard
			variant={variant}
			finish={_minutes === 0 && _seconds === 0}
			running={running}
		>
			{' '}
			<p>{variant}</p>
			{_minutes > 0 || _seconds > 0 ? (
				<StyledSpan>
					{_minutes.toString().padStart(2, '0')}:{_seconds.toString().padStart(2, '0')}
				</StyledSpan>
			) : (
				'00:00'
			)}
			{variant === 'Exercise' ? (
				<PlayIcon />
			) : variant === 'Pause' ? (
				<PauseIcon />
			) : (
				<SetPauseIcon />
			)}
		</StyledTimerCard>
	);
}
`````
